pkg/test/local/pilot.go

```diff
@@ -0,0 +1,178 @@
+	"strconv"
+	"istio.io/istio/pilot/pkg/networking/core/v1alpha3"
+	"istio.io/istio/pilot/pkg/networking/plugin/registry"
+	"net/http"
+	"net"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	defaultGrpcMaxStreams = 100000
+)
+
+type args struct {
+	Namespace string
+	Mesh *meshconfig.MeshConfig
+	ServiceControllers *aggregate.Controller
+	ConfigDir string
+	MixerServiceAccountName []string
+	DiscoveryServiceOptions envoyv1.DiscoveryServiceOptions
+}
+
+type Pilot struct {
+	discoveryOptions envoyv1.DiscoveryServiceOptions
+	configController model.ConfigStoreCache
+	configMonitor *configmonitor.Monitor
+	discoveryService *envoyv1.DiscoveryService
+	xdsServer   *envoyv2.DiscoveryServer
+	grpcServer *grpc.Server
+	httpServer *http.Server
+}
+
+func NewPilot(args args) (*Pilot, error) {
+	// TODO(nmittler): Better place for this?
+	envoyv1.ValidateClusters = false
+
+	configController := configController()
+	configMonitor := fileMonitor(args.ConfigDir, configController)
+
+	environment := model.Environment{
+		Mesh:             args.Mesh,
+		IstioConfigStore: model.MakeIstioStore(configController),
+		ServiceDiscovery: args.ServiceControllers,
+		ServiceAccounts:  args.ServiceControllers,
+		MixerSAN:         args.MixerServiceAccountName,
+	}
+
+	// Set up discovery service
+	discovery, err := envoyv1.NewDiscoveryService(
+		args.ServiceControllers,
+		configController,
+		environment,
+		args.DiscoveryServiceOptions,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create discovery service: %v", err)
+	}
+
+	// For now we create the gRPC server sourcing data from Pilot's older data model.
+	grpcServer := grpcServer()
+	xdsServer := envoyv2.NewDiscoveryServer(grpcServer, environment, v1alpha3.NewConfigGenerator(registry.NewPlugins()))
+	xdsServer.InitDebug(discovery.RestContainer.ServeMux, args.ServiceControllers)
+
+	// TODO: decouple v2 from the cache invalidation, use direct listeners.
+	envoyv1.V2ClearCache = xdsServer.ClearCacheFunc()
+
+	httpServer := &http.Server{
+		Addr:    ":" + strconv.Itoa(args.DiscoveryServiceOptions.Port),
+		Handler: discovery.RestContainer,
+	}
+
+	// TODO(nmittler): Add monitoring port
+
+	return &Pilot {
+		discoveryOptions: args.DiscoveryServiceOptions,
+		configController: configController,
+		configMonitor: configMonitor,
+		discoveryService: discovery,
+		xdsServer: xdsServer,
+		grpcServer: grpcServer,
+		httpServer: httpServer,
+	}, nil
+}
+
+func (p *Pilot) Start(stop chan struct{}) error {
+	// Start the HTTP server.
+	addr := p.httpServer.Addr
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		return err
+	}
+
+	// Start the GRPC server.
+	grpcListener, err := net.Listen("tcp", p.discoveryOptions.GrpcAddr)
+	if err != nil {
+		return err
+	}
+
+	// Now run all the servers.
+	go p.configController.Run(stop)
+	p.configMonitor.Start(stop)
+	go func() {
+		if err = p.httpServer.Serve(listener); err != nil {
+			log.Warna(err)
+		}
+	}()
+	go func() {
```

Comment:
This is also duplicating a lot of stuff in bootstrap - why not use bootstrap directly ?